src/flag_gems/ops/cross_entropy_loss.py

```diff
@@ -39,12 +39,12 @@ def celoss_indice_kernel(
     tgt_mask = offset_d < D
     tgt = tl.load(tgt_ptrs, mask=tgt_mask, other=0)
 
-    ignore_mask = not (tgt == ignore_index)
+    ignore_mask = not (tgt == ignore_index) and tgt_mask
 
-    w_ptrs = w_ptr + tgt
-    w_tgt = tl.load(w_ptrs, mask=tgt_mask, other=0).to(tl.float32)
-    w_tgt_ptrs = w_tgt_ptr + pid_n * D + offset_d
-    tl.store(w_tgt_ptrs, w_tgt, mask=tgt_mask and ignore_mask)
+    w_tgt = tl.load(w_ptr + tgt, mask=ignore_mask, other=0).to(tl.float32)
+    if reduction == 1:
+        w_tgt_sum = tl.sum(w_tgt)
+        tl.atomic_add(w_tgt_ptr, w_tgt_sum)
```

Comment:
We can also sum the weight of each example in another kernel(sum_and_scale) to avoid using `atomic_add`. Atomic operations may hurt performance but we need some comparison to verify that.

Reply:
Or we can gather the weight for each example and write it out to an ((N,D) tensor. Then in sum_and_scale we can load them more effectively.

Reply:
done